### PR TITLE
[issue-6147] [BE] fix: replace FINAL with LIMIT 1 BY on thread-by-id span aggregation

### PR DESCRIPTION
**Pull request by siddiqueirshad**
## Details

Clicking a thread still times out on large `spans` tables because `SELECT_TRACES_THREAD_BY_ID` was the one remaining ThreadDAO query that aggregated with `FROM spans final`. List/count/stats already switched to a `spans_deduped` CTE (`ORDER BY ... LIMIT 1 BY id`) so ClickHouse can apply `workspace_id` / `project_id` / `trace_id` filters before ReplacingMergeTree dedup. This change uses that same shape on the retrieve-by-id path, which is what `POST /traces/threads/retrieve` runs.

PR #6358 fixed the listing queries. PR #6295 tried to convert all four `spans_agg` CTEs, including this one, but was closed after those listing changes landed. The by-id query was left on `FINAL`.

## Change checklist
- [x] User facing
- [ ] Documentation update

## Issues

- Resolves #6147

## AI-WATERMARK

AI-WATERMARK: yes

- If yes:
  - Tools: Cursor
  - Model(s): Cursor Grok 4.6
  - Scope: root-cause investigation of #6147, ThreadDAO SQL change, and ThreadDAOImplTest regression
  - Human verification: confirmed the leftover `FROM spans final` is only in `SELECT_TRACES_THREAD_BY_ID`; matched the existing list-query `spans_deduped` pattern; ran `ThreadDAOImplTest` (12 tests, 0 failures)

## Testing

- Command: `mvn -B test -Dtest=ThreadDAOImplTest` in `apps/opik-backend` (JDK 25 Maven container)
- Result: Tests run: 12, Failures: 0, Errors: 0, Skipped: 0; BUILD SUCCESS
- Scenarios: new regression asserts the by-id query no longer contains `FROM spans final`, dedups via `LIMIT 1 BY id`, and keeps the `trace_id IN (SELECT DISTINCT id FROM traces_ids)` filter
- Existing ThreadDAOImplTest cases (firstThreadOrEmpty, traces_final_ids prefilter gate, thread_id pushdown) still pass
- ClickHouse integration tests / FindTraceThreadsResourceTest were not run locally (no Java/Maven on the host besides Docker; no live ClickHouse for resource tests)

## Documentation

No documentation update. Behavior is unchanged for callers; only the ClickHouse scan for thread-by-id retrieve is aligned with the existing list/count/stats queries.

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ThreadDAO.java [hunk 1]
```diff
@@ -781,7 +781,8 @@ thread_model_id NOT IN (SELECT entity_id FROM fsc)
      *  - The creator of the thread, which is the created_by of the first trace in the list.
      *  - The creation time of the thread, which is the created_at of the first trace in the list.
      ***/
-    private static final String SELECT_TRACES_THREAD_BY_ID = """
+    @VisibleForTesting
+    static final String SELECT_TRACES_THREAD_BY_ID = """
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ThreadDAO.java:785]: **[high] The dedup behavior change is covered only by substring assertions on the SQL constant, so a regression that double-counts span usage or picks a stale span version still passes**

SELECT_TRACES_THREAD_BY_ID was widened from private to package-private purely so ThreadDAOImplTest.ThreadByIdSpansDedup can assert that the string contains 'LIMIT 1 BY id' and does not contain 'FROM spans final'. That test never executes the query: it cannot show that a span with two ReplacingMergeTree versions (the PARTIAL_INSERT-then-INSERT case that motivates the change) now contributes its usage/total_estimated_cost exactly once, that the winning version is the one with the greatest last_updated_at, or that providers is not duplicated. A rewrite that keeps the literal text but breaks the semantics (e.g. moving the trace_id predicate after the dedup, or dropping the last_updated_at tie-break) would still pass every assertion, and the whole behavior the PR claims to fix is unverified at the ClickHouse boundary.

💡 Add a ThreadDAO integration test (the existing ClickHouse-backed thread resource/DAO test suites) that inserts a span, then a second version of that span with a different parent_span_id and a later last_updated_at, calls findById, and asserts the returned TraceThread usage, total_estimated_cost and providers equal the single-version values.

_rule: `backend-testing-test-the-real-failure-boundary-not-a-mock-of-it`_

File: apps/opik-backend/src/test/java/com/comet/opik/domain/ThreadDAOImplTest.java [hunk 2]
```diff
@@ -213,4 +213,36 @@ private static String traceThreadsFinalCte(String sql) {
...
     }
+
+    @Nested
+    @DisplayName("thread-by-id spans aggregation")
+    class ThreadByIdSpansDedup {
+
+        @Test
+        @DisplayName("SELECT_TRACES_THREAD_BY_ID dedups spans with LIMIT 1 BY id instead of FINAL")
+        void threadByIdDoesNotUseSpansFinal() {
+            String sql = ThreadDAOImpl.SELECT_TRACES_THREAD_BY_ID;
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/domain/ThreadDAOImplTest.java:224]: **[medium] The FINAL-to-LIMIT-1-BY dedup change is covered only by substring matching on the SQL constant, with no test that executes the query**

Replacing `FROM spans final` with a spans_deduped CTE (`ORDER BY (workspace_id, project_id, trace_id, id) DESC, last_updated_at DESC LIMIT 1 BY id`) changes result semantics on a ReplacingMergeTree: picking the wrong version, or failing to collapse versions, silently double-counts sumMap(usage) / sum(total_estimated_cost) and inflates the providers array for a thread whose spans were updated. This test only greps the constant, so it cannot falsify any of that. The only integration coverage of this code path, TracesResourceTest.GetTraceThread#getTraceThread (line 6038), builds its spans with usage(null), provider(null) and totalEstimatedCost(null) and never updates a span, so no test in the repo would fail if the new dedup picked the stale span version or emitted duplicates.

💡 Add an integration test on the thread-retrieve endpoint that creates spans with non-null usage/provider/total_estimated_cost, then re-sends the same span ids with changed usage/cost (producing a second ReplacingMergeTree version), and asserts the returned TraceThread usage, total_estimated_cost and providers equal the latest-version values exactly — not double the sum.

_rule: `backend-testing-test-load-bearing-sql-gating-logic-flags-substring-matches-c`_

File: apps/opik-backend/src/test/java/com/comet/opik/domain/ThreadDAOImplTest.java [hunk 3]
```diff
@@ -213,4 +213,36 @@ private static String traceThreadsFinalCte(String sql) {
...
+
+            assertThat(sql).doesNotContain("FROM spans final");
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/domain/ThreadDAOImplTest.java]: **[medium] The anti-FINAL guard matches only the lowercase spelling, so a reintroduced `FROM spans FINAL` passes**

doesNotContain("FROM spans final") is a case-sensitive, whitespace-exact literal. Both spellings are in active use in the DAO layer (for `FROM spans` specifically: 16 lowercase, 8 uppercase) — `FROM spans FINAL` appears verbatim in KpiCardDAO.java (202, 312, 512), TraceDAO.java (1261, 2360, 2790), DatasetItemVersionDAO.java:2557 and SpanMetricsQueries.java:120 — so the single most likely regression (someone re-adding FINAL in the house style, or writing `FROM spans  final` / `FROM spans\n FINAL`) slips through this guard green. The detector does not assert the invariant it claims to protect (no FINAL scan on the spans read of SELECT_TRACES_THREAD_BY_ID); it asserts one arbitrary spelling of it.

💡 Assert the invariant case-insensitively and spelling-tolerantly, e.g. assertThat(sql).doesNotContainIgnoringCase("spans final") or match a regex such as Pattern.compile("(?i)\\bFROM\\s+spans\\s+FINAL\\b") against the constant, and apply the same check to the spans_agg slice at line 237.

_rule: `backend-correctness-guards-lints-precondition-checks-and-detectors-must-assert-t`_

**Review by andrescrz — COMMENTED** [review]
## Opik reviewer (mined from your team's review history)

4 findings — 1 high · 2 medium · 1 low. Suppressed by team conventions: see `suppressed.md`.

React 👍/👎 on each comment — your feedback helps tune what it flags.

File: apps/opik-backend/src/test/java/com/comet/opik/domain/ThreadDAOImplTest.java [hunk 4]
```diff
@@ -213,4 +213,36 @@ private static String traceThreadsFinalCte(String sql) {
...
+
+            String spansDeduped = cte(sql, "spans_deduped AS (", "), spans_agg AS (");
+            assertThat(spansDeduped).contains("FROM spans");
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/domain/ThreadDAOImplTest.java]: **[low] assertThat(spansDeduped).contains("FROM spans") is satisfied by the very SQL the test is meant to reject**

"FROM spans" is a prefix of both "FROM spans FINAL"/"FROM spans final" and "FROM spans_deduped", so this assertion holds for every variant the test class exists to distinguish. Inside the spans_deduped slice nothing actually pins that the read is a plain, non-FINAL table scan — that responsibility rests entirely on the case-sensitive whole-SQL check at line 226.

💡 Pin the exact token boundary inside the CTE, e.g. assertThat(spansDeduped).containsPattern("FROM\\s+spans\\s*\\n") or assert the full line "FROM spans\n                WHERE workspace_id = :workspace_id", plus a case-insensitive doesNotContainIgnoringCase("spans final") on the slice.

_rule: `backend-testing-make-the-assertion-actually-falsify-the-change-under-test`_

**Comment by siddiqueirshad** [comment]
Addressed the test-guard findings:

- Anti-`FINAL` assertions are now case-insensitive (`(?i)\bFROM\s+spans\s+FINAL\b`) on the whole query and both CTE slices.
- The `spans_deduped` plain-table read is pinned with an exact token-boundary pattern (`FROM\s+spans\s*\n`) instead of the `contains("FROM spans")` prefix check.

The broader ClickHouse integration tests (two-version span dedup / endpoint-level double-count checks) still need a backend test environment, which isn't available here, so I left those for follow-up rather than adding unverified integration coverage.

**Comment by siddiqueirshad** [comment]
Hi @thiagohora, when you get a chance, could you please take a look at this PR? Thanks!
